📡 refactor: MCP Runtime Config Sync with Redis Distributed Locking

atefbellaaj-at-slalom opened the pull request:
## Summary

This PR introduces lazy-loading configs in the registry for private/shared MCP servers with distributed locking, efficient batch operations, and runtime configuration reload capability. The new architecture separates private server lifecycle management from the registry, enabling zero-downtime updates to MCP server configurations, runtime server access control, and tier migrations without requiring application restarts.

  

## Key Changes

  

### 1. Timestamp-Based Staleness Detection

The core mechanism uses simple timestamp comparison to detect when connections need to be recreated:

  

**MCPConnection Timestamp Tracking** (Already exists from previous work)

* Connection has `createdAt` property set once at creation

* `isStale(configUpdatedAt)` method for staleness detection

* Simple comparison: `connection.createdAt < config.lastUpdatedAt` = stale

  

**Type System Update** (`packages/api/src/mcp/types/index.ts:156`)

* Renamed `cachedAt` → `lastUpdatedAt` for clearer semantics

* All cached configs now include `lastUpdatedAt` timestamp

* Timestamp automatically set on cache operations

* Propagates across Redis cluster automatically

  

### 2. Lazy Connection Invalidation

Connections are validated before reuse with automatic staleness detection and recreation upon retrieval:

  

**ConnectionsRepository** (`packages/api/src/mcp/ConnectionsRepository.ts:43-51`)

```ts

// Check if config was cached/updated since connection was created

if (serverConfig.lastUpdatedAt && existingConnection.isStale(serverConfig.lastUpdatedAt)) {

logger.info(`Existing connection for ${serverName} is outdated. Recreating...`, {

connectionCreated: new Date(existingConnection.createdAt).toISOString(),

configCachedAt: new Date(serverConfig.lastUpdatedAt).toISOString(),

});

await existingConnection.disconnect();

this.connections.delete(serverName);

// Fall through to create new connection

}

```

  

**UserConnectionManager** (`packages/api/src/mcp/UserConnectionManager.ts:82`)

```ts

// Check if config was updated since connection was created

if (!config || (config.lastUpdatedAt && connection.isStale(config.lastUpdatedAt))) {

logger.info(`Config was updated, disconnecting stale connection`);

await this.disconnectUserConnection(userId, serverName);

connection = undefined; // Force new connection creation

}

```

  

### 3. Leader-Only Mode Adjustment

**Registry Cache Configuration** (`packages/api/src/mcp/registry/MCPServersRegistry.ts:24-32`)

  

Changed shared caches from leader-only to allow runtime updates:

  

```ts

// Before: Leader-only (blocks runtime updates on followers)

public readonly sharedAppServers = ServerConfigsCacheFactory.create('App', true);

public readonly sharedUserServers = ServerConfigsCacheFactory.create('User', true);

  

// After: Runtime updates enabled

public readonly sharedAppServers = ServerConfigsCacheFactory.create('App', false);

public readonly sharedUserServers = ServerConfigsCacheFactory.create('User', false);

```

  

**Rationale:**

* Initializer still coordinates leadership for startup

* Runtime updates bypass leader check for responsiveness across all nodes

  

### 4. MCPPrivateServerLoader - Runtime Server Management

**New component** (`packages/api/src/mcp/registry/MCPPrivateServerLoader.ts`) handles all private server lifecycle operations:

  

* **`loadPrivateServers()`** (`MCPPrivateServerLoader.ts:32-110`)

- Lazy-loads private servers on-demand with distributed locking

- Prevents partial loads and race conditions across multiple app instances

- TTL-synchronized load flags prevent cache desync issues

- Handles edge cases: process crashes, cache eviction, users with 0 servers

  

* **`updatePrivateServer()`** (`MCPPrivateServerLoader.ts:120-126`)

- Propagates metadata changes to all affected users efficiently

- Uses pattern-based Redis SCAN for targeted updates

- Sets fresh `lastUpdatedAt` timestamp triggering lazy connection invalidation

  

* **`updatePrivateServerAccess()`** (`MCPPrivateServerLoader.ts:136-169`)

- Handles permission changes (grant/revoke access)

- Compares current vs desired state and applies minimal changes

  

* **Tier Migration Methods** (`MCPPrivateServerLoader.ts:171-236`)

- `promoteToSharedServer()`: Migrates private → shared (PUBLIC)

- `demoteToPrivateServer()`: Migrates shared → private

- Ensures no duplication across cache tiers

  

### 5. Specialized Private Server Cache Layer

**New abstraction** for managing per-user server configurations with efficient batch operations:

  

* **`PrivateServerConfigsCacheBase`** (`cache/PrivateServerConfigs/PrivateServerConfigsCacheBase.ts`)

- Abstract base defining batch operation primitives

- No business logic - pure cache orchestration

  

* **`PrivateServerConfigsCacheRedis`** (`cache/PrivateServerConfigs/PrivateServerConfigsCacheRedis.ts`)

- Redis implementation with pattern-based SCAN operations

- Efficient batch updates using Redis pipelining

- Automatically sets `lastUpdatedAt` timestamp on all operations

- Methods: `updateServerConfigIfExists()`, `addServerConfigIfCacheExists()`, `removeServerConfigIfCacheExists()`, `findUsersWithServer()`

  

* **`PrivateServerConfigsCacheInMemory`** (`cache/PrivateServerConfigs/PrivateServerConfigsCacheInMemory.ts`)

- In-memory fallback for single-node deployments

  

### 6. Load Status Cache with Distributed Locking

**New cache** (`cache/PrivateServersLoadStatusCache.ts`) solves three critical issues:

  

1. **TTL Synchronization**: Load flags expire in sync with cache entries

2. **Race Condition Prevention**: Distributed locks prevent concurrent loads

3. **Cache Eviction Detection**: When cache expires, load flag expires too

  

**Key Methods:**

* `acquireLoadLock()` / `releaseLoadLock()`: Distributed locking

* `waitForLoad()`: Cooperative waiting mechanism

* `setLoaded()` / `isLoaded()`: TTL-synchronized load tracking

  

### 7. Registry Simplification

**MCPServersRegistry** (`packages/api/src/mcp/registry/MCPServersRegistry.ts:23-125`) simplified to focus on unified query interface:

**Changes:**

* Replaced `Map<string, ServerConfigsCache>` with dedicated `PrivateServerConfigsCache`

* Removed private server CRUD methods (moved to cache layer)

* Renamed `resetServer()` → `removeServer()` with clearer semantics

* Removed `getPrivateServerConfig()` (use unified `getServerConfig()` instead)

  

### 8. Initializer Simplification

**MCPServersInitializer** (`packages/api/src/mcp/registry/MCPServersInitializer.ts`) refactored to focus only on startup initialization:

  

**Removed Methods:**

* `reInitializeServer()` - Runtime updates now handled by MCPPrivateServerLoader

* `initPrivateServers()` - Lazy-loading now handled by MCPPrivateServerLoader

* `ensurePrivateServerInitialized()` - Logic moved to loader

  

**Exposed:**

* `initializeServer()`: Now public for reuse by loader when needed

  

### 9. MCPManager API Cleanup

**MCPManager** (`packages/api/src/mcp/MCPManager.ts`) simplified to focus on connection management:

  

**Removed:**

* `updateConfig()` - Use MCPPrivateServerLoader methods instead

* `initPrivateServers()` - Use MCPPrivateServerLoader.loadPrivateServers() instead

  

## Technical Details


  

### Batch Operation Efficiency

**Pattern-Based Updates** (`PrivateServerConfigsCacheRedis.ts:25-73`):

```ts

// Efficient: Single SCAN finds all affected users

const pattern = `*${PREFIX}::*:${serverName}`;

for await (const key of redis.scanIterator({ MATCH: pattern })) {

keysToUpdate.push(key);

}

  

// Batch update using Redis pipeline with fresh timestamp

const updatedConfig = { ...config, lastUpdatedAt: Date.now() };

const multi = redis.multi();

for (const key of keysToUpdate) {

multi.set(key, updatedConfig, { XX: true });

}

await multi.exec();

```

  

### TTL Synchronization

Load status flags share the same TTL as cache entries, preventing scenarios where:

- Cache expires but flag still says "loaded" → would serve empty results

- Flag expires but cache still has data → unnecessary DB queries

  

### Distributed Locking Mechanism

**Lock Acquisition Flow** (`PrivateServersLoadStatusCache.ts:67-86`):

```ts

// Process A: Tries to acquire lock

const lockAcquired = await loadStatusCache.acquireLoadLock(userId);

if (!lockAcquired) {

// Process B: Wait for Process A to finish

const completed = await loadStatusCache.waitForLoad(userId, 5000);

if (!completed) {

// Process A crashed - retry lock acquisition

const retryLock = await loadStatusCache.acquireLoadLock(userId);

}

}

```

  

**Edge Cases Handled:**

* Process crashes mid-load: Lock TTL expires (30s), other processes retry

* Multiple processes start simultaneously: Only one acquires lock, others wait

* Lock holder completes: Sets loaded flag, waiting processes detect and proceed

  

### Server Tier Migration Self-Healing

When a server moves between tiers (e.g., App → User or User → App), the system automatically routes to the correct tier:

  

**How it works:**

```ts

// MCPManager.getConnection() routing logic

if (await this.appConnections!.has(args.serverName)) {

// has() checks sharedAppServers cache

// If config moved to user tier, cache returns false

// Automatically falls through to user connection branch

return this.appConnections!.get(args.serverName);

} else if (args.user?.id) {

// Creates/gets user connection with config from correct tier

return this.getUserConnection(args);

}

```

  

**Orphaned Connection Handling:**

* Orphaned connections (from tier migration) remain in memory

* Cleaned up automatically by existing idle timeout mechanism

* `has()` disconnects orphaned connections when config not found

  
## Reset cache issue 

The standardCache().clear() method was not respecting namespace boundaries due to a Keyv bug workaround that sets namespace and keyPrefixSeparator after Keyv instantiation. This caused cache.clear() to reset the entire Redis cache instead of just the specific namespace, breaking test isolation and potentially affecting production multi-tenant cache operations.

Impact:
  - Test "should clear all servers for a specific user" was failing in cluster mode
  - Calling reset() on a user-specific cache would delete data for ALL users
  - Issue #10487 

### Solution and Changes Made

1. Namespace-Aware clear() Override (Primary Fix)

Added a custom clear() implementation in cacheFactory.ts that:
  - Uses Redis SCAN with namespace-specific patterns (globalPrefix::namespace:*)
  - Deletes keys in parallel chunks for efficiency
  - Handles both cluster mode (individual DEL) and single-node mode (batch DEL) to avoid CROSSSLOT errors
  - Only affects keys within the target namespace

2. Test Cleanup Improvements (Secondary Fix)

Updated all 6 MCP integration test files to use parallel deletion in afterEach hooks instead of sequential deletion, reducing race conditions in cluster mode:
  - PrivateServerConfigsCacheRedis.cache_integration.spec.ts
  - ServerConfigsCacheRedis.cache_integration.spec.ts
  - MCPServersRegistry.cache_integration.spec.ts
  - MCPServersInitializer.cache_integration.spec.ts
  - PrivateServersLoadStatusCache.cache_integration.spec.ts
  - RegistryStatusCache.cache_integration.spec.ts

3. Re-enabled Reset Call
Uncommented the reset() call in MCPPrivateServerLoader.loadPrivateServers() that was previously disabled due to this bug.
4. moved leaderOnly setter to the constructor of the BaseConfigCache.ts and reset() method checks if the flag is set before deciding to perform a leader check. 

## Change Type

* [x] New feature (non-breaking change which adds functionality)

* [x] Refactor (improves code structure and maintainability)

  

## Testing

* Comprehensive unit tests for MCPPrivateServerLoader (555+ lines)

* Integration tests against actual Redis for batch operations (373+ lines)

* Cache integration tests for load status and distributed locking (418+ lines)

* All existing MCP tests updated for `lastUpdatedAt` rename

* All tests passing

  

## Checklist

* [x] My code adheres to this project's style guidelines

* [x] I have performed a self-review of my own code

* [x] I have commented in any complex areas of my code

* [x] My changes do not introduce new warnings

* [x] I have written tests demonstrating that my changes are effective or that my feature works

* [x] Local unit tests pass with my changes

* [x] Any changes dependent on mine have been merged and published in downstream modules

  

---

  

Atef Bellaaj slalom.bellaaj@external.daimlertruck.com on behalf of Daimler Truck AG.

Copyright (c) 2025 Daimler Truck AG

nhtruong commented:
I actually had a very different idea of how to implement this so that a Connection Repository is always given the most up-to-date config when it creates a new connection: It no longer stores the configs themselves. Instead, whenever it needs to create a new connection, it fetches the config through the MCPServersRegistry, the source of truth, by providing a serverName. I believe that would be a lot simpler than figuring out which config is stale since configs retrieved from the registry is always the latest.

atefbellaaj-at-slalom commented:
> I actually had a very different idea of how to implement this so that a Connection Repository is always given the most up-to-date config when it creates a new connection: It no longer stores the configs themselves. Instead, whenever it needs to create a new connection, it fetches the config through the MCPServersRegistry, the source of truth, by providing a serverName. I believe that would be a lot simpler than figuring out which config is stale since configs retrieved from the registry is always the latest.

@nhtruong 

Thanks for reviewing this. I might be missing a detail, but do we actually recreate a connection on every request? If the goal is to reuse existing connections (when possible) and keep them in memory, we still need a way to verify whether a cached connection is compatible with the latest config. In that scenario, the things that become stale after a config update are the connections, not the configs themselves.

I’m open to refactoring if you have a concrete flow in mind, and I’d be interested in any specifics on how the repository would know when an in-memory connection should be discarded.

If this isn’t critical right now, though, I’d prefer to push forward on the remaining pieces needed to finish the “Add MCP from UI” feature.

nhtruong commented:
@atefbellaaj-at-slalom 

Regarding reusing connections, we had a [discussion](https://github.com/danny-avila/LibreChat/discussions/8790#discussioncomment-13959953) on this topic a few months ago. The TLDR is that we will eventually want user connections to be ephemeral, or at least give the admin to determine the Connection TTL for each server. Right now user connections have a fixed 15 min TTL across the board. In the future, each server can have their own TTL in their server config (With 0 TTL means ephemeral and Infinity means permanent). You may have noticed that the logic that handles TTL logic for user connections currently resides in MCPManager, which it's the responsibility of the ConnectionsRepositry class. Moving that logic into the repository is one of the future refactoring when we implement custom TTL. However that's beyond the scope of this PR.

Regarding connections with stale server configs, I'd like to propose a simpler solution:
- When a config is added/updated via the registry, we will also update its `lastUpdatedAt` timestamp
- When a connection is created, we will also note its `createdAt`. This timestamp can either be a prop of the MCPConnection, or the MCPConnectionsRepository will have to keep track of it since it's responsible for returning the correct and update-to-date connections)
- When the repository is asked to return a connection, it will compare those 2 timestamps to determine if it can return the existing connection as is or create a new one. Note that the repo will access the newest server config via the registry, our authoritative on Server Configs, instead of keeping tracking of its own server configs cache. That's my primary concern for the current approach in the PR: the repository has its own logic to retrieve the server configs instead of via the registry. That is, we now open doors to let different components of the app to have their own definition of what the "correct" config for a certain server is. This will lead to inconsistency and very hard to fix bugs in the future.

nhtruong commented:
```ts
public async addSharedServer(serverName: string, config: t.ParsedServerConfig): Promise<void> {
  if (config.requiresOAuth || config.startup === false) {
    await this.sharedUserServers.add(serverName, config);
  } else {
    await this.sharedAppServers.add(serverName, config);
  }
}
```

I strongly believe that whether a server is App-level or User-level should be explicitly stated instead of via the side effect of whether a server should be initialized during startup. When an admin adds a new  App-level server config via the UI for example, the frontend logic should not have to be aware of this awkward implementation detail. So instead of sending back `startup: true`, it should send back something like `scope: sharedAppServers` to register the config. The controller on the backend then will use this `scope` value to deterministically pick the correct collection in the registry to add the config to i.e. `if (config.scope === `sharedAppServers`) registry.sharedAppServers.add(serverName, config);`

In the same vein, we should get rid of `config.startup` because it's very confusing, esp when we can now add app-level severs on the fly AFTER startup. So this side-effect of `startup` no longer determines whether a server is app-level or user-level. This is another reason (the other being confusing/cryptic) why we should avoiding using side effects. This refactoring is worth its own PR and out of the scope of this PR tho.

nhtruong commented:
Here's  the implementation of `MCPSeversRegistry.getServerConfig`
```ts
  public async getServerConfig(
    serverName: string,
    userId?: string,
  ): Promise<t.ParsedServerConfig | undefined> {
    const sharedAppServer = await this.sharedAppServers.get(serverName);
    if (sharedAppServer) return sharedAppServer;

    const sharedUserServer = await this.sharedUserServers.get(serverName);
    if (sharedUserServer) return sharedUserServer;

    const privateUserServer = await this.privateUserServers.get(userId)?.get(serverName);
    if (privateUserServer) return privateUserServer;

    return undefined;
  }
```

Basically this func gives you a different answer depending on the scope. If you don't provide it with a userId, it returns a sharedAppServer or null. In retrospect, I should have named the param `ownerId` instead of `userId` as in the future, we will also have servers on the `sharedGroupServers` scope as well. But basically, depending "who's asking", this function will return the appropriate server for their scope.

So, the Repository class should now be aware of the scope of servers whose connections it's managing. We can add a `ownerId` prop that can be set in the constructor (a null value means it's off the sharedAppServers scope).

This implementation of `MCPSeversRegistry.getServerConfig` is not ideal because it doesn't handle the case where 2 servers of the same name existing in different scopes. However, right now we're operating on the assumption that the server names are always unique since the UI right now only shows the server names without anyway of telling which scope a server belongs to. Hence this func was created to cater to the existing functionality: you only need the serverName to determine which server is being worked on. In a future iteration, we should update both the Frontend/Backend to be scope-aware.

nhtruong commented on packages/api/src/mcp/registry/MCPServersRegistry.ts:
```diff
@@ -56,17 +53,33 @@ class MCPServersRegistry {
     return undefined;
   }
 
+  public async getPrivateServerConfig(
```
Is the this function only created for testing?

nhtruong commented on packages/api/src/mcp/registry/MCPServersInitializer.ts:
```diff
@@ -56,7 +58,62 @@ export class MCPServersInitializer {
+    const { serverName, config, user, isPrivateServer } = args;
+    await registry.resetServer(serverName, user?.id);
+    if (isPrivateServer) {
+      if (!user || !user.id) {
```
(Nit) `if (!user?.id)` has the same effect :D

nhtruong commented on packages/api/src/mcp/registry/MCPServersInitializer.ts:
```diff
@@ -56,7 +58,62 @@ export class MCPServersInitializer {
+          'User must be provided for private server updates',
+        );
+      }
+      await registry.addPrivateUserServer(user.id, serverName, config);
```
Did you mean to call `initPrivateServers` here instead?

nhtruong commented on packages/api/src/mcp/ConnectionsRepository.ts:
```diff
@@ -1,37 +1,66 @@
+    if (existingConnection) {
+      // Check if config was cached/updated since connection was created
+      if (serverConfig.cachedAt && existingConnection.isStale(serverConfig.cachedAt)) {
+        logger.info(`${this.prefix(serverName)} Config updated, reconnecting`, {
```
Nit: If I see this log message, I would think it comes from the bit of code that updates the config. maybe "Existing connection for {servername} is outdated. Recreating a new connection."

atefbellaaj-at-slalom replied:
Yes that is a clearer log message. thanks.

nhtruong commented:
@atefbellaaj-at-slalom looking great over all and thank you for putting in the work. I just have some small comments here and there :)

atefbellaaj-at-slalom commented on packages/api/src/mcp/MCPManager.ts:
```diff
@@ -41,8 +41,21 @@ export class MCPManager extends UserConnectionManager {
+    this.appConnections = new ConnectionsRepository(undefined);
+  }
+  /** update  config of an mcp server with possibility to change Server Tier based on whether it is private and or its configuration*/
+  public async updateConfig(args: {
```
I guess you are right. I don't know why i had the impression that MCPManager was the entry point of the mcp package. maybe we should rename the MCPManager to MCPConnectionManager in the future. 
I will remove these wrappers.

nhtruong replied:
MCPManager was the monolith catch-all for everything MCP related. It was very bloated and lacked separation of concerns (i.e. very hard to maintain and use). One of my overhauls moved a lot logic out of the manager (which gave birth to the registry, repository, factory...).

atefbellaaj-at-slalom commented:
I am converting this back to Draft to make some changes.

nhtruong commented on packages/api/src/mcp/registry/MCPServersInitializer.ts:
```diff
@@ -56,7 +58,62 @@ export class MCPServersInitializer {
+  }): Promise<void> {
+    const { serverName, config, user, isPrivateServer } = args;
+    await registry.resetServer(serverName, user?.id);
+    if (isPrivateServer) {
```
That's fair. I think instead of userId, we can use ownerId, if there's no owner, it's shared and not private, which is very intuitive. OwnerId also future-proofs this and allows us to expand the concept of private servers to groups/orgs...

atefbellaaj-at-slalom commented on packages/api/src/mcp/registry/MCPServersInitializer.ts:
```diff
@@ -56,7 +58,62 @@ export class MCPServersInitializer {
+  public static async reInitializeServer(args: {
+    serverName: string;
+    config: t.MCPOptions;
+    user?: TUser;
```
method removed in favor of the new design

atefbellaaj-at-slalom commented on packages/api/src/mcp/MCPManager.ts:
```diff
@@ -41,8 +41,21 @@ export class MCPManager extends UserConnectionManager {
+  }
+
+  /** ensure the registry is updated with private servers */
+  public async initPrivateServers(userId: string, configs: t.MCPServers) {
```
removed

atefbellaaj-at-slalom commented on packages/api/src/mcp/registry/MCPServersRegistry.ts:
```diff
@@ -86,6 +116,46 @@ class MCPServersRegistry {
     this.privateUserServers.clear();
   }
+
+  public async resetServer(serverName: string, userId?: string): Promise<void> {
```
done

Commit:
🔄 Refactoring: MCP Runtime Configuration Reload
 - PrivateServerConfigs own cache classes (inMemory and Redis).
 - Connections staleness detection by comparing (connection.createdAt and config.LastUpdatedAt)
 - ConnectionsRepo access Registry instead of in memory config dict and renew stale connections
 - MCPManager: adjusted init of ConnectionsRepo (app level)
 - UserConnectionManager: renew stale connections
 - skipped test, to test "should only clear keys in its own namespace"
 - MCPPrivateServerLoader: new component to manage logic of loading / editing private servers on runtime
 - PrivateServersLoadStatusCache to track private server cache status
 - New unit and integration tests.
Misc:
 - add es lint rule to enforce line between class methods

Commit:
Fix cluster mode batch update and delete workarround. Fixed unit tests for cluster mode.

Commit:
Fix Keyv redis clear cache namespace  awareness issue + Integration tests fixes

Commit:
chore: address copilot comments

Commit:
Fixing rebase issue: removed the mcp config fallback in single getServerConfig method:
- to not to interfere with the logic of the right Tier (APP/USER/Private)
- If userId is null, the getServerConfig should not return configs that are a SharedUser tier and not APP tier

Commit:
chore: add dev-staging branch to workflow triggers for backend, cache integration, and ESLint checks
